pkg/server/bootstrap_server.go

```diff
@@ -0,0 +1,122 @@
+	"github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// ensure bootstrapServer implements the
+// Server interface.
+var _ = Server(&bootstrapServer{})
+
+type bootstrapServer struct {
+
+	// serverFileSystem is the file system used
+	// for reading files on the server.
+	serverFileSystem string
+
+	// serverBaseDir is the root, relative to which
+	// the machine pool, configs will be picked
+	serverBaseDir string
+
+	// serverKubeConfPath is the path on the local machine from
+	// where the kubeconfig should be read.
+	serverKubeConfPath string
+
+	// machineFileSystem defines the filesystem to be used
+	// for storing the files created by the server.
+	// eg: kubeconfig, node-annotations.
+	machineFileSystem string
+}
+
+// NewBootstrapServer initializes a new Bootstrap server that implements
+// the Server interface.
+func NewBootstrapServer(sfs, dir, mfs, kubeConf string) (Server, error) {
+	return &bootstrapServer{
+		serverFileSystem:   sfs,
+		serverBaseDir:      dir,
+		machineFileSystem:  mfs,
+		serverKubeConfPath: kubeConf,
+	}, nil
+}
+
+// GetConfig fetches the machine config(type - Ignition) from the bootstrap server,
+// based on the pool request.
+// It returns nil for conf, error if the config isn't found. It returns a formatted
+// error if any other error is encountered during its operations.
+//
+// The method does the following:
+//
+// 1. Read the machine config pool by using the following path template:
+// 		"/<serverFileSystem>/<serverBaseDir>/<machineConfiPoolName>.yaml"
+//
+// 2. Read the currentConfig field from the Status and read the config file
+//    using the following path template:
+// 		"/<serverFileSystem>/<serverBaseDir>/<machineConfiPoolName>/configs/<currentConfig>.yaml"
+//
+// 3. Load the machine config.
+// 4. Execute the etcd template function based on the etcd_index, if passed in the request .
+// 5. Append the machine annotations file.
+// 6. Append the KubeConfig file.
+func (bsc *bootstrapServer) GetConfig(cr poolRequest) (*ignv2_2types.Config, error) {
+
+	// 1. Read the Machine Config Pool object.
+	fileName := path.Join(bsc.serverFileSystem, bsc.serverBaseDir, cr.machinePool+".yaml")
+	data, err := ioutil.ReadFile(fileName)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("server: could not read file %s, err: %v", fileName, err)
+	}
+
+	mp := new(v1.MachineConfigPool)
+	err = yaml.Unmarshal(data, mp)
+	if err != nil {
+		return nil, fmt.Errorf("server: could not unmarshal file %s, err: %v", fileName, err)
+	}
+
+	currConf := mp.Status.CurrentMachineConfig
+
+	// 2. Read the Machine Config object.
+	fileName = path.Join(bsc.serverBaseDir, cr.machinePool, "configs", currConf+".yaml")
+	data, err = ioutil.ReadFile(fileName)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("server: could not unmarshal file %s, err: %v", fileName, err)
```

Comment:
I believe this is pre unmarshaling for section 2. `"server: could not read file %s, err: %v"` may be more appropriate.